subsys/net/lib/lwm2m/lwm2m_obj_connmon.c

```diff
@@ -0,0 +1,397 @@
+
+/* Connectivity Monitoring resource IDs */
+#define CONNMON_NETWORK_BEARER_ID		0
+#define CONNMON_AVAIL_NETWORK_BEARER_ID		1
+#define CONNMON_RADIO_SIGNAL_STRENGTH		2
+#define CONNMON_LINK_QUALITY			3
+#define CONNMON_IP_ADDRESSES			4
+#define CONNMON_ROUTER_IP_ADDRESSES		5
+#define CONNMON_LINK_UTILIZATION		6
+#define CONNMON_APN				7
+#define CONNMON_CELLID				8
+#define CONNMON_SMNC				9
+#define CONNMON_SMCC				10
+
+#define CONNMON_MAX_ID				11
+
+#define CONNMON_STRING_SHORT			8
+
+#if defined(CONFIG_LWM2M_CONN_MON_BEARER_MAX)
+#define CONNMON_AVAIL_BEARER_MAX	CONFIG_LWM2M_CONN_MON_BEARER_MAX
+#else
+#define CONNMON_AVAIL_BEARER_MAX	5
+#endif
+
+#if defined(CONFIG_LWM2M_CONN_MON_APN_MAX)
+#define CONNMON_APN_MAX			CONFIG_LWM2M_CONN_MON_APN_MAX
+#else
+#define CONNMON_APN_MAX			5
+#endif
+
+#if defined(CONFIG_NET_IF_MAX_IPV6_COUNT) \
+	&& defined(CONFIG_NET_IF_MAX_IPV6_COUNT)
+#define CONNMON_IP_ADDRESS_MAX		(CONFIG_NET_IF_MAX_IPV6_COUNT + \
+					CONFIG_NET_IF_MAX_IPV4_COUNT)
+#elif defined(CONFIG_NET_IF_MAX_IPV6_COUNT)
+#define CONNMON_IP_ADDRESS_MAX		CONFIG_NET_IF_MAX_IPV6_COUNT
+#elif defined(CONFIG_NET_IF_MAX_IPV4_COUNT)
+#define CONNMON_IP_ADDRESS_MAX		CONFIG_NET_IF_MAX_IPV4_COUNT
+#else
+#define CONNMON_IP_ADDRESS_MAX		1
+#endif
+
+#if defined(CONFIG_NET_MAX_ROUTERS)
+#define CONNMON_ROUTER_IP_ADDRESS_MAX	CONFIG_NET_MAX_ROUTERS
+#else
+#define CONNMON_ROUTER_IP_ADDRESS_MAX	1
+#endif
+
+/* resource state variables */
+static s8_t net_bearer;
+static s8_t avail_net_bearer[CONNMON_AVAIL_BEARER_MAX];
+static s8_t rss;
+static u8_t link_quality;
+static char ip_address[CONNMON_IP_ADDRESS_MAX][INET6_ADDRSTRLEN];
+static char router_ip_address[CONNMON_ROUTER_IP_ADDRESS_MAX][INET6_ADDRSTRLEN];
+static char apn[CONNMON_APN_MAX][CONNMON_STRING_SHORT];
+static u16_t mnc;
+static u16_t mcc;
+
+static u8_t bearer_count;
+static u8_t ip_address_count;
+static u8_t router_ip_address_count;
+static u8_t apn_count;
+
+/* only 1 instance of Connection Monitoring object exists */
+static struct lwm2m_engine_obj connmon;
+static struct lwm2m_engine_obj_field fields[] = {
+	OBJ_FIELD_DATA(CONNMON_NETWORK_BEARER_ID, R, U8),
+	OBJ_FIELD(CONNMON_AVAIL_NETWORK_BEARER_ID, R, U8,
+		  CONNMON_AVAIL_BEARER_MAX),
+	OBJ_FIELD_DATA(CONNMON_RADIO_SIGNAL_STRENGTH, R, S8),
+	OBJ_FIELD_DATA(CONNMON_LINK_QUALITY, R, U8),
+	OBJ_FIELD(CONNMON_IP_ADDRESSES, R, STRING, CONNMON_IP_ADDRESS_MAX),
+	OBJ_FIELD(CONNMON_ROUTER_IP_ADDRESSES, R_OPT, STRING,
+		  CONFIG_NET_MAX_ROUTERS),
+	OBJ_FIELD_DATA(CONNMON_LINK_UTILIZATION, R_OPT, U8),
+	OBJ_FIELD(CONNMON_APN, R_OPT, STRING, CONNMON_APN_MAX),
+	OBJ_FIELD_DATA(CONNMON_CELLID, R_OPT, S32),
+	OBJ_FIELD_DATA(CONNMON_SMNC, R_OPT, U16),
+	OBJ_FIELD_DATA(CONNMON_SMCC, R_OPT, U16)
+};
+
+static struct lwm2m_engine_obj_inst inst;
+static struct lwm2m_engine_res_inst res[CONNMON_MAX_ID];
+
+/* special setter functions */
+
+int lwm2m_connmon_add_bearer(u8_t bearer_id)
+{
+	int index;
+
+	if (bearer_id >= 50) {
```

Comment:
It would be more clearer if we check against a define instead of magic value, now it is very unclear what this special value means.

Reply:
Most of the odd-ball logic was removed from the connectivity monitoring object, so this issue goes away.  It will need to be re-addressed if we later integrate with the networking subsystem somehow.